File: src/main/java/com/handwoong/everyonewaiter/service/StoreServiceImpl.java
package com.handwoong.everyonewaiter.service;

import static com.handwoong.everyonewaiter.enums.ErrorCode.MEMBER_NOT_FOUND;
import static com.handwoong.everyonewaiter.enums.ErrorCode.STORE_NOT_FOUND;
import static com.handwoong.everyonewaiter.enums.ErrorCode.TELEPHONE_NUMBER_EXISTS;

import com.handwoong.everyonewaiter.domain.Member;
import com.handwoong.everyonewaiter.domain.Store;
import com.handwoong.everyonewaiter.dto.StoreDto;
import com.handwoong.everyonewaiter.exception.CustomException;
import com.handwoong.everyonewaiter.repository.MemberRepository;
import com.handwoong.everyonewaiter.repository.StoreRepository;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class StoreServiceImpl implements StoreService {

    private final StoreRepository storeRepository;

    private final MemberRepository memberRepository;

    @Override
    @Transactional
    public Long register(String username, StoreDto.RequestDto storeDto) {
        Member findMember = findMemberByUsername(username);
        existsTelephone(username, storeDto.getTelephoneNumber());

        Store store = Store.createStore(storeDto, findMember);
        storeRepository.save(store);

        return store.getId();
    }

    @Override
    @Transactional
    public void update(String username, Long storeId, StoreDto.RequestDto storeDto) {
        existsTelephone(username, storeDto.getTelephoneNumber());
        Store store = findMemberStore(username, storeId);
        store.updateStore(storeDto);
    }

    @Override
    @Transactional
    public void delete(String username, Long storeId) {
        Store store = findMemberStore(username, storeId);
        storeRepository.delete(store);
    }

    @Override
    public List<StoreDto.ResponseDto> findStoreList(String username) {
        Member findMember = findMemberByUsername(username);
        List<Store> storeList = storeRepository.findAllByMemberId(findMember.getId());

        return storeList.stream()
                .map(StoreDto.ResponseDto::from)
                .toList();
    }

    @Override
    public StoreDto.ResponseDto findStore(String username, Long storeId) {
        Store store = findMemberStore(username, storeId);
        return StoreDto.ResponseDto.from(store);
    }

    private Member findMemberByUsername(String username) {
        return memberRepository.findByUsername(username)
                .orElseThrow(() -> new CustomException(MEMBER_NOT_FOUND));
    }

    private Store findMemberStore(String username, Long storeId) {
        return storeRepository.findMemberStore(username, storeId)
                .orElseThrow(() -> new CustomException(STORE_NOT_FOUND));
    }

    private void existsTelephone(String username, String telephoneNumber) {
        boolean existTelNumber = storeRepository.existTelNumber(username, telephoneNumber);
        if (existTelNumber) {
            throw new CustomException(TELEPHONE_NUMBER_EXISTS);
        }
    }
}

File: src/main/kotlin/com/handwoong/everyonewaiter/service/store/StoreServiceImpl.kt
package com.handwoong.everyonewaiter.service.store

import com.handwoong.everyonewaiter.domain.member.Member
import com.handwoong.everyonewaiter.domain.store.Store
import com.handwoong.everyonewaiter.dto.store.StoreRequest
import com.handwoong.everyonewaiter.dto.store.StoreResponse
import com.handwoong.everyonewaiter.exception.ErrorCode.*
import com.handwoong.everyonewaiter.repository.member.MemberRepository
import com.handwoong.everyonewaiter.repository.store.StoreRepository
import com.handwoong.everyonewaiter.util.throwFail
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class StoreServiceImpl(
    private val memberRepository: MemberRepository,
    private val storeRepository: StoreRepository,
) : StoreService {

    @Transactional
    override fun register(username: String, storeDto: StoreRequest) {
        val findMember = findMemberByUsername(username)
        existsTelephone(
            username = username,
            telephone = storeDto.telephone,
        )
        val createStore = Store.createStore(storeDto, findMember)
        storeRepository.save(createStore)
    }

    @Transactional
    override fun update(username: String, storeId: Long, storeDto: StoreRequest) {
        val findStore = findMemberStore(storeId, username)
        existsTelephone(
            username = username,
            telephone = storeDto.telephone,
        )
        findStore.update(storeDto)
    }

    @Transactional
    override fun delete(username: String, storeId: Long) {
        val findStore = findMemberStore(storeId, username)
        storeRepository.delete(findStore)
    }

    override fun findStore(username: String, storeId: Long): StoreResponse {
        val findStore = findMemberStore(storeId, username)
        return StoreResponse.of(findStore)
    }

    override fun findStoreList(username: String): List<StoreResponse> {
        return storeRepository.findAllStore(username)
            .map(StoreResponse::of)
    }

    private fun findMemberByUsername(username: String): Member {
        return memberRepository.findMember(username) ?: throwFail(MEMBER_NOT_FOUND)
    }

    private fun findMemberStore(storeId: Long, username: String): Store {
        return storeRepository.findStore(storeId, username) ?: throwFail(STORE_NOT_FOUND)
    }

    private fun existsTelephone(username: String, telephone: String) {
        if (storeRepository.existsStore(username, telephone)) {
            throwFail(TELEPHONE_EXISTS)
        }
    }

}
